distributed/profile.py

```diff
@@ -59,21 +60,40 @@ def identifier(frame):
         )
 
 
+def _f_lineno(frame):
+    f_lineno = frame.f_lineno
+    if f_lineno is not None:
+        return f_lineno
+
+    f_lasti = frame.f_lasti
+    code = frame.f_code
+    prev_line = code.co_firstlineno
+
+    for start, next_line in dis.findlinestarts(code):
+        if f_lasti < start:
+            return prev_line
+        prev_line = next_line
+
+    return prev_line
+
+
 def repr_frame(frame):
     """Render a frame as a line for inclusion into a text traceback"""
     co = frame.f_code
-    text = f'  File "{co.co_filename}", line {frame.f_lineno}, in {co.co_name}'
-    line = linecache.getline(co.co_filename, frame.f_lineno, frame.f_globals).lstrip()
+    f_lineno = _f_lineno(frame)
+    text = f'  File "{co.co_filename}", line {f_lineno}, in {co.co_name}'
+    line = linecache.getline(co.co_filename, f_lineno, frame.f_globals).lstrip()
     return text + "\n\t" + line
 
 
 def info_frame(frame):
     co = frame.f_code
-    line = linecache.getline(co.co_filename, frame.f_lineno, frame.f_globals).lstrip()
+    f_lineno = _f_lineno(frame)
```

Comment:
frames can have a `frame.f_lineno=None` In Python 3.10 this seems to be much more likely eg when dispatching a JUMP_ABSOLUTE: https://github.com/python/cpython/commit/f5e97b72fecff9b9ced35704ec5e6cad29e2825d#diff-ebc983d9f91e5bcf73500e377ac65e85863c4f77fd5b6b6caf4fcdf7c0f0b057R2777

we can calculate a nearby line with `frame.f_lasti` and `dis.findlinestarts`

Reply:
a smaller reproducer for this issue is here: https://gist.github.com/graingert/f5e81264dd3c3ab8a705809096fcc283#file-demo-py-L27


related to the logger gaining ~thousand DequeHandlers over the course of the test suite https://github.com/dask/distributed/issues/5973